p2p/src/handshake.rs

```diff
@@ -0,0 +1,309 @@
+        ) -> Result<GetKey<T, K, E>, crate::Error> {
+            #[allow(clippy::expect_used)]
+            let write_half = connection
+                .write
+                .as_mut()
+                .expect("Never fails as in this function we already have the stream.");
+            write_half.as_ref().writable().await?;
+
+            // We take our public key from our `id` and will replace it with theirs when we read it
+            // Packing length and message in one network packet for efficiency
+            let data = id.public_key.encode();
+
+            let data = &crypto.encrypt(data)?;
+
+            let mut buf = Vec::<u8>::with_capacity(data.len() + 1);
+            #[allow(clippy::cast_possible_truncation)]
+            buf.push(data.len() as u8);
+            buf.extend_from_slice(data.as_slice());
+
+            write_half.write_all(&buf).await?;
+            Ok(GetKey(id, broker, connection, crypto))
+        }
+    }
+
+    /// Peer that needs to get key.
+    pub struct GetKey<T, K, E>(peer::Id, Broker, Connection, Cryptographer<T, K, E>)
+    where
+        T: Encode + Decode + Send + Clone + 'static,
+        K: KeyExchangeScheme + Send + 'static,
+        E: Encryptor + Send + 'static;
+
+    impl<T, K, E> GetKey<T, K, E>
+    where
+        T: Encode + Decode + Send + Clone + 'static,
+        K: KeyExchangeScheme + Send + 'static,
+        E: Encryptor + Send + 'static,
+    {
+        /// Read the peer's public key
+        ///
+        /// # Panics
+        /// Never
+        pub async fn read_their_public_key(
+            Self(mut id, broker, mut connection, crypto): Self,
+        ) -> Result<Ready<T, K, E>, crate::Error> {
+            #[allow(clippy::unwrap_used)]
+            let read_half = connection.read.as_mut().unwrap();
+            let size = read_half.read_u8().await? as usize;
+            if size >= crate::peer::MAX_HANDSHAKE_LENGTH {
+                return Err(crate::HandshakeError::Length(size).into());
+            }
+            // Reading public key
+            read_half.as_ref().readable().await?;
+            let mut data = vec![0_u8; size];
+            let _ = read_half.read_exact(&mut data).await?;
+
+            let data = crypto.decrypt(data)?;
+
+            let pub_key = Decode::decode(&mut data.as_slice())?;
+
+            id.public_key = pub_key;
+            Ok(Ready(id, broker, connection, crypto))
+        }
+    }
+
+    /// Peer that is ready for communication after finishing the
+    /// handshake process.
+    pub struct Ready<T, K, E>(peer::Id, pub Broker, pub Connection, Cryptographer<T, K, E>)
+    where
+        T: Encode + Decode + Send + Clone + 'static,
+        K: KeyExchangeScheme + Send + 'static,
+        E: Encryptor + Send + 'static;
+
+    /// Peer in disconnected state.
+    #[derive(Debug)]
+    pub struct Disconnected(peer::Id);
+
+    /// Peer in broken state.
+    #[derive(Debug)]
+    pub struct Broken(peer::Id, crate::Error);
```

Comment:
should I care to mention that you're implementing `Identifiable` by hand? You could have done this and have it done for you:

```suggestion
    pub struct Broken(peer, crate::Error);
```

Reply:
Temporarily. As you see from the draft status this PR is WIP

Reply:
ah, pardon me

Reply:
Do we need an issue for this?

Reply:
#2620